internal/controller/connectionsecret/requestname_extractor.go

```diff
@@ -0,0 +1,292 @@
+
+	// Index lookups
+	ErrNoPairedResourcesFound = errors.New("no AtlasDeployment and no AtlasDatabaseUser found")
+	ErrNoDeploymentFound      = errors.New("no AtlasDeployment found")
+	ErrManyDeployments        = errors.New("multiple AtlasDeployments found")
+	ErrNoUserFound            = errors.New("no AtlasDatabaseUser found")
+	ErrManyUsers              = errors.New("multiple AtlasDatabaseUsers found")
+)
+
+// ConnSecretIdentifiers holds the values extracted from a reconcile request name.
+type ConnSecretIdentifiers struct {
+	ProjectID        string
+	ProjectName      string
+	ClusterName      string
+	DatabaseUsername string
+}
+
+// ConnSecretPair represents the pairing of an AtlasDeployment and an AtlasDatabaseUser
+// required to construct a ConnectionSecret. It holds resolved identifiers and the corresponding resources.
+// NOTE: this struct intentionally stores only ProjectID (not all identifiers) to keep only necessary information.
+type ConnSecretPair struct {
+	ProjectID  string
+	Deployment *akov2.AtlasDeployment
+	User       *akov2.AtlasDatabaseUser
+}
+
+// ConnectionData contains all connection information required to populate
+// the Kubernetes Secret, including standard and SRV URLs and optional Private Link URLs.
+type ConnSecretData struct {
+	DBUserName      string
+	Password        string
+	ConnURL         string
+	SrvConnURL      string
+	PrivateConnURLs []PrivateLinkConnURLs
+}
+
+// PrivateLinkConnURLs holds all Private Link connection strings for a single endpoint set.
+// Multiple entries allow for multiple private link configurations per deployment.
+type PrivateLinkConnURLs struct {
+	PvtConnURL      string
+	PvtSrvConnURL   string
+	PvtShardConnURL string
+}
+
+// CreateK8sFormat returns the Secret name in the Kubernetes naming format: <projectName>-<clusterName>-<username>
+func CreateK8sFormat(projectName string, clusterName string, databaseUsername string) string {
+	return strings.Join([]string{
+		kube.NormalizeIdentifier(projectName),
+		kube.NormalizeIdentifier(clusterName),
+		kube.NormalizeIdentifier(databaseUsername),
+	}, "-")
+}
+
+// CreateInternalFormat returns the Secret name in the internal format used by watchers: <projectID>$<clusterName>$<username>
+func CreateInternalFormat(projectID string, clusterName string, databaseUsername string) string {
+	return strings.Join([]string{
+		projectID,
+		kube.NormalizeIdentifier(clusterName),
+		kube.NormalizeIdentifier(databaseUsername),
+	}, InternalSeparator)
+}
+
+// LoadRequestIdentifiers determines whether the request name is internal or K8s format
+// and extracts ProjectID, ClusterName, and DatabaseUsername.
+func LoadRequestIdentifiers(ctx context.Context, c client.Client, req types.NamespacedName) (ConnSecretIdentifiers, error) {
+	var ids ConnSecretIdentifiers
+
+	// === Internal format: <ProjectID>$<ClusterName>$<DatabaseUserName>
+	if strings.Contains(req.Name, InternalSeparator) {
+		parts := strings.SplitN(req.Name, InternalSeparator, 3)
```

Comment:
If the number of part is not 3, something is wrong, not matter if less or more items.
```suggestion
		parts := strings.Split(req.Name, InternalSeparator)
```